File: src/Benchmark/CI.hs
{-|
Module      : Benchmark.CI
Description : CI environment integration for benchmark results
Stability   : experimental

Detects CI environment and formats regression output for GitLab and GitHub Actions pipelines.
Writes markdown reports for merge request artifacts and job summaries.
-}
module Benchmark.CI
  ( CIMode (..)
  , detectCIMode
  , formatForCI
  , writeArtifactReport
  , writeGitHubStepSummary
  ) where

import Benchmark.Report.Markdown (markdownRegressionReport)
import Benchmark.Types (MetricRegression (..), RegressionResult (..))
import Control.Monad (forM_)
import Data.Text qualified as T
import Data.Text.IO qualified as TIO
import Data.Time.Clock.POSIX (getPOSIXTime)
import System.Environment (lookupEnv)
import Text.Printf (printf)

-- | CI environment detection.
data CIMode = GitLab | GitHub | None
  deriving (Show, Eq)

-- | Detect if running in a known CI environment.
detectCIMode :: IO CIMode
detectCIMode = do
  gitlab <- lookupEnv "GITLAB_CI"
  github <- lookupEnv "GITHUB_ACTIONS"
  return $ case (gitlab, github) of
    (Just "true", _) -> GitLab
    (_, Just "true") -> GitHub
    _ -> None

{-| Format regression results for CI output.
In GitLab, section markers and colored output help visibility.
GitHub Actions uses plain stdout (no section markers).
-}
formatForCI :: CIMode -> RegressionResult -> IO ()
formatForCI None _ = return ()
formatForCI GitLab result = do
  epoch <- (round <$> getPOSIXTime) :: IO Int
  putStrLn ""
  putStrLn $ "\x1b[0Ksection_start:" ++ show epoch ++ ":benchmark_results[collapsed=false]"
  putStrLn "\x1b[0K\x1b[1;34m========== Benchmark Regression Check ==========\x1b[0m"
  putStrLn ""

  forM_ (regressionMetrics result) $ \m -> do
    let status :: String
        status = if metricRegressed m then "\x1b[1;31mREGRESSED\x1b[0m" else "\x1b[1;32mok\x1b[0m"
    printf
      "  %s: %.2fms -> %.2fms (%+.1f%%) [%s]\n"
      (T.unpack $ metricName m)
      (metricBaseline m)
      (metricCurrent m)
      (metricChange m * 100)
      status

  putStrLn ""
  if regressionPassed result
    then putStrLn "\x1b[1;32m[PASS] Benchmark passed - no regressions detected\x1b[0m"
    else putStrLn "\x1b[1;31m[FAIL] Benchmark FAILED - regression detected\x1b[0m"

  putStrLn ""
  putStrLn $ "\x1b[0Ksection_end:" ++ show epoch ++ ":benchmark_results"
formatForCI GitHub result = do
  putStrLn ""
  putStrLn "========== Benchmark Regression Check =========="
  putStrLn ""

  forM_ (regressionMetrics result) $ \m -> do
    let status = if metricRegressed m then "REGRESSED" else "ok" :: String
    printf
      "  %s: %.2fms -> %.2fms (%+.1f%%) [%s]\n"
      (T.unpack $ metricName m)
      (metricBaseline m)
      (metricCurrent m)
      (metricChange m * 100)
      status

  putStrLn ""
  if regressionPassed result
    then putStrLn "[PASS] Benchmark passed - no regressions detected"
    else putStrLn "[FAIL] Benchmark FAILED - regression detected"
  putStrLn ""

-- | Write a markdown report for CI artifacts / GitHub Step Summary.
writeArtifactReport :: FilePath -> RegressionResult -> IO ()
writeArtifactReport path result =
  TIO.writeFile path (markdownRegressionReport result)

-- | Write the markdown regression report to $GITHUB_STEP_SUMMARY if set.
writeGitHubStepSummary :: RegressionResult -> IO ()
writeGitHubStepSummary result = do
  mpath <- lookupEnv "GITHUB_STEP_SUMMARY"
  case mpath of
    Nothing -> return ()
    Just path -> TIO.appendFile path (markdownRegressionReport result)

File: crates/gauntlet-report/src/ci.rs
//! CI integration: regression output shaped for the detected CI provider, plus
//! the markdown artifact those providers collect.
//!
//! Only regressions are reported here. A CI run's job is to answer "did this
//! change make things slower"; the full benchmark detail belongs in the
//! markdown/HTML artifacts.

use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

use async_trait::async_trait;

use crate::error::Result;
use crate::format::{ms, signed_pct};
use crate::markdown;
use crate::{RegressionResult, Reporter};

/// The CI provider we appear to be running under.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CiMode {
    GitLab,
    GitHub,
    None,
}

/// Detect the CI provider from the environment.
///
/// GitLab is checked first: a GitLab job can shell out in ways that leave
/// `GITHUB_ACTIONS` set, but not the reverse.
pub fn detect_ci() -> CiMode {
    let is_true = |var: &str| std::env::var(var).map(|v| v == "true").unwrap_or(false);

    if is_true("GITLAB_CI") {
        CiMode::GitLab
    } else if is_true("GITHUB_ACTIONS") {
        CiMode::GitHub
    } else {
        CiMode::None
    }
}

/// Render the regression check for a CI job log.
///
/// GitLab gets collapsible section markers and colour; GitHub Actions renders
/// its own log formatting and gets plain text.
pub fn ci_log(mode: CiMode, result: &RegressionResult, epoch_secs: u64) -> String {
    /// Wraps a status word in colour, or leaves it alone.
    type Colorize = fn(&str) -> String;

    let (color_bad, color_good): (Colorize, Colorize) = match mode {
        CiMode::GitLab => (
            |s| format!("\x1b[1;31m{s}\x1b[0m"),
            |s| format!("\x1b[1;32m{s}\x1b[0m"),
        ),
        _ => (|s| s.to_string(), |s| s.to_string()),
    };

    let mut out = String::from("\n");
    if mode == CiMode::GitLab {
        out.push_str(&format!(
            "\x1b[0Ksection_start:{epoch_secs}:benchmark_results[collapsed=false]\n"
        ));
        out.push_str(
            "\x1b[0K\x1b[1;34m========== Benchmark Regression Check ==========\x1b[0m\n\n",
        );
    } else {
        out.push_str("========== Benchmark Regression Check ==========\n\n");
    }

    for m in &result.metrics {
        let status = if m.regressed {
            color_bad("REGRESSED")
        } else {
            color_good("ok")
        };
        out.push_str(&format!(
            "  {}: {}ms -> {}ms ({}) [{}]\n",
            m.name,
            ms(m.baseline),
            ms(m.current),
            signed_pct(m.change),
            status,
        ));
    }

    out.push('\n');
    out.push_str(&if result.passed {
        color_good("[PASS] Benchmark passed - no regressions detected")
    } else {
        color_bad("[FAIL] Benchmark FAILED - regression detected")
    });
    out.push('\n');

    if mode == CiMode::GitLab {
        out.push_str(&format!(
            "\x1b[0Ksection_end:{epoch_secs}:benchmark_results\n"
        ));
    }
    out
}

/// Emits CI-shaped regression output, writes a markdown artifact, and on GitHub
/// appends to the job's Step Summary. A no-op outside CI.
pub struct CiReporter {
    mode: CiMode,
    artifact_dir: PathBuf,
}

impl CiReporter {
    /// Build a reporter for the detected provider, writing artifacts to `dir`.
    pub fn detect(artifact_dir: impl Into<PathBuf>) -> Self {
        Self {
            mode: detect_ci(),
            artifact_dir: artifact_dir.into(),
        }
    }

    pub fn with_mode(mode: CiMode, artifact_dir: impl Into<PathBuf>) -> Self {
        Self {
            mode,
            artifact_dir: artifact_dir.into(),
        }
    }

    /// True when there is no CI provider to report to.
    pub fn is_inactive(&self) -> bool {
        self.mode == CiMode::None
    }
}

#[async_trait]
impl Reporter for CiReporter {
    async fn on_regression(&self, result: &RegressionResult) -> Result<()> {
        if self.mode == CiMode::None {
            return Ok(());
        }

        let epoch = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);

        print!("{}", ci_log(self.mode, result, epoch));

        let report = markdown::regression_report(result);
        let artifact = self
            .artifact_dir
            .join(format!("benchmark-report-{epoch}.md"));
        markdown::write_new(&artifact, &report)?;

        // GitHub renders this file at the top of the job's summary page.
        if self.mode == CiMode::GitHub {
            if let Some(path) = std::env::var_os("GITHUB_STEP_SUMMARY") {
                markdown::append(std::path::Path::new(&path), &report)?;
            }
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::baseline::MetricRegression;

    fn result(passed: bool) -> RegressionResult {
        RegressionResult {
            baseline: "main".into(),
            metrics: vec![MetricRegression {
                name: "mean".into(),
                baseline: 10.0,
                current: 12.0,
                change: 0.2,
                threshold: 0.10,
                regressed: !passed,
            }],
            passed,
        }
    }

    #[test]
    fn gitlab_output_is_wrapped_in_matching_section_markers() {
        let text = ci_log(CiMode::GitLab, &result(false), 1_700_000_000);
        assert!(text.contains("section_start:1700000000:benchmark_results"));
        assert!(text.contains("section_end:1700000000:benchmark_results"));
        assert!(text.contains('\x1b'), "GitLab renders ANSI colour");
    }

    #[test]
    fn github_output_is_plain_and_unsectioned() {
        let text = ci_log(CiMode::GitHub, &result(false), 1_700_000_000);
        assert!(!text.contains("section_start"));
        assert!(!text.contains('\x1b'));
        assert!(text.contains("========== Benchmark Regression Check =========="));
    }

    #[test]
    fn metric_lines_carry_the_full_before_after_delta() {
        let text = ci_log(CiMode::GitHub, &result(false), 0);
        assert!(text.contains("  mean: 10.00ms -> 12.00ms (+20.0%) [REGRESSED]"));
        assert!(text.contains("[FAIL] Benchmark FAILED - regression detected"));
    }

    #[test]
    fn passing_runs_say_pass() {
        let text = ci_log(CiMode::GitHub, &result(true), 0);
        assert!(text.contains("[ok]"));
        assert!(text.contains("[PASS] Benchmark passed - no regressions detected"));
    }

    #[tokio::test]
    async fn outside_ci_the_reporter_writes_nothing() {
        let dir = std::env::temp_dir().join("gauntlet-ci-inactive-test");
        let _ = std::fs::remove_dir_all(&dir);

        let reporter = CiReporter::with_mode(CiMode::None, &dir);
        assert!(reporter.is_inactive());
        reporter.on_regression(&result(false)).await.unwrap();

        assert!(!dir.exists(), "no artifact directory is created");
    }

    #[tokio::test]
    async fn in_ci_a_markdown_artifact_is_written() {
        let dir =
            std::env::temp_dir().join(format!("gauntlet-ci-artifact-test-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);

        CiReporter::with_mode(CiMode::GitLab, &dir)
            .on_regression(&result(false))
            .await
            .unwrap();

        let written: Vec<_> = std::fs::read_dir(&dir).unwrap().flatten().collect();
        assert_eq!(written.len(), 1);
        let contents = std::fs::read_to_string(written[0].path()).unwrap();
        assert!(contents.contains("## Regression Check vs `main`"));

        let _ = std::fs::remove_dir_all(&dir);
    }
}
